src/Type/SubtractedType.php

```diff
@@ -0,0 +1,313 @@
+use PHPStan\Reflection\PropertyReflection;
+use PHPStan\TrinaryLogic;
+use PHPStan\Type\Generic\TemplateTypeMap;
+use PHPStan\Type\Generic\TemplateTypeVariance;
+
+class SubtractedType implements Type, CompoundType
+{
+	private $originalType;
+	private $subtractedType;
+
+	private function __construct(Type $originalType, Type $subtractedType)
+	{
+		$this->originalType = $originalType;
+		$this->subtractedType = $subtractedType;
+	}
+
+	/**
+	 * TODO: move/use logic from TypeCombinator here, to create the type
+	 * @see TypeCombinator::remove()
+	 * The goal should be to normalize the type if possible (subtract a class from a union type of classes etc)
+	 * Only if that is not possible use this type, e.g. when subtracting a constant string from the general string type
+	 */
+	public static function create(Type $originalType, Type $subtractedType): Type
+	{
+		if ($subtractedType->isSuperTypeOf($originalType)->yes() || $originalType instanceof NeverType) {
+			return new NeverType();
+		}
+
+		if ($originalType->accepts($subtractedType, false)->no()) {
+			return $originalType;
+		}
+
+		return new self($originalType, $subtractedType);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getReferencedClasses(): array
+	{
+		return array_merge(
+			$this->originalType->getReferencedClasses(),
+			$this->subtractedType->getReferencedClasses()
+		);
+	}
+
+	public function accepts(Type $type, bool $strictTypes): TrinaryLogic
+	{
+		$original = $this->originalType->accepts($type, $strictTypes);
+		if (!$original->yes()) {
+			return $original;
+		}
+
+		return TrinaryLogic::createFromBoolean(!$this->subtractedType->equals($type));
+	}
+
+	public function isSuperTypeOf(Type $type): TrinaryLogic
+	{
+		// TODO: Implement isSuperTypeOf() method.
+	}
+
+	public function equals(Type $type): bool
+	{
+		if (!$type instanceof self) {
+			return false;
+		}
+
+		return $this->originalType->equals($type) && !$this->subtractedType->equals($type);
+	}
+
+	public function describe(VerbosityLevel $level): string
+	{
+		return $this->originalType->describe($level) . '~' . $this->subtractedType->describe($level);
+	}
+
+	public function canAccessProperties(): TrinaryLogic
+	{
+		if ($this->subtractedType->equals(new ObjectWithoutClassType())) {
+			return TrinaryLogic::createNo();
+		}
+
+		if ($this->originalType instanceof MixedType) {
+			return TrinaryLogic::createYes();
+		}
+
+		$original = $this->originalType->canAccessProperties();
+		if (!$original->yes()) {
+			return $original;
+		}
+
+		if ($this->subtractedType->canAccessProperties()->no()) {
+			return TrinaryLogic::createYes();
+		}
+
+		/*
+		 * TODO:
+		 * Both the original and the subtracted type can access properties
+		 * now we need to determine if *only* the subtracted type has it, or if it is
+		 * present without the subtracted type as well.
+		 */
+		return TrinaryLogic::createMaybe();
+	}
+
+	public function hasProperty(string $propertyName): TrinaryLogic
```

Comment:
What do you think of this logic?

If both have the property we return a maybe, as we can't be sure where it came from.

Note that at this point i'm not 100% sure if it is even possible to subtract one class from another if we would 'normalize' union types during creation.